// @process-engine-js/data_model_contracts/dist/commonjs/entity_collection_reference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class EntityCollectionReference {
    constructor(namespace, type, id, relatedType, relatedAttribute, resolve) {
        this._namespace = undefined;
        this._type = undefined;
        this._relatedType = undefined;
        this._relatedAttribute = undefined;
        this._id = undefined;
        this.resolve = false;
        this._namespace = namespace;
        this._type = type;
        this._id = id;
        this._relatedType = relatedType;
        this._relatedAttribute = relatedAttribute;
        if (resolve) {
            this.resolve = resolve;
        }
    }
    get namespace() {
        return this._namespace;
    }
    get type() {
        return this._type;
    }
    get id() {
        return this._id;
    }
    get relatedType() {
        return this._relatedType;
    }
    get relatedAttribute() {
        return this._relatedAttribute;
    }
    toPojo() {
        const data = {
            id: this.id,
            _meta: {
                isCollectionRef: true,
                type: this.type,
                namespace: this.namespace,
                relatedType: this.relatedType,
                relatedAttribute: this.relatedAttribute
            }
        };
        return data;
    }
}
exports.EntityCollectionReference = EntityCollectionReference;
