// @process-engine-js/core_contracts/dist/commonjs/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityDiscoveryTag = 'entityDiscoveryTag';
exports.RouterDiscoveryTag = 'routerDiscoveryTag';
exports.ExtensionDiscoveryTag = 'extensionDiscoveryTag';
exports.TokenAdapterDiscoveryTag = 'tokenAdapterDiscoveryTag';
exports.FeatureGeneratorDiscoveryTag = 'featureGeneratorDiscoveryTag';
var TokenType;
(function (TokenType) {
    TokenType[TokenType["jwt"] = 0] = "jwt";
})(TokenType = exports.TokenType || (exports.TokenType = {}));
var SaveMode;
(function (SaveMode) {
    SaveMode[SaveMode["virtual"] = 0] = "virtual";
    SaveMode[SaveMode["persistance"] = 1] = "persistance";
    SaveMode[SaveMode["add"] = 2] = "add";
})(SaveMode = exports.SaveMode || (exports.SaveMode = {}));
var QueryMethod;
(function (QueryMethod) {
    QueryMethod[QueryMethod["findOne"] = 0] = "findOne";
    QueryMethod[QueryMethod["count"] = 1] = "count";
    QueryMethod[QueryMethod["all"] = 2] = "all";
})(QueryMethod = exports.QueryMethod || (exports.QueryMethod = {}));
var LogicalOperatorType;
(function (LogicalOperatorType) {
    LogicalOperatorType[LogicalOperatorType["and"] = 0] = "and";
    LogicalOperatorType[LogicalOperatorType["or"] = 1] = "or";
})(LogicalOperatorType = exports.LogicalOperatorType || (exports.LogicalOperatorType = {}));
var SchemaAttributeType;
(function (SchemaAttributeType) {
    SchemaAttributeType[SchemaAttributeType["string"] = 0] = "string";
    SchemaAttributeType[SchemaAttributeType["object"] = 1] = "object";
    SchemaAttributeType[SchemaAttributeType["number"] = 2] = "number";
    SchemaAttributeType[SchemaAttributeType["boolean"] = 3] = "boolean";
    SchemaAttributeType[SchemaAttributeType["int"] = 4] = "int";
    SchemaAttributeType[SchemaAttributeType["Date"] = 5] = "Date";
    SchemaAttributeType[SchemaAttributeType["Blob"] = 6] = "Blob";
})(SchemaAttributeType = exports.SchemaAttributeType || (exports.SchemaAttributeType = {}));
exports.RegistrationKeySpacer = ':';
exports.DefaultNamespace = undefined;
var ClaimActionType;
(function (ClaimActionType) {
    ClaimActionType[ClaimActionType["create"] = 0] = "create";
    ClaimActionType[ClaimActionType["read"] = 1] = "read";
    ClaimActionType[ClaimActionType["write"] = 2] = "write";
    ClaimActionType[ClaimActionType["delete"] = 3] = "delete";
    ClaimActionType[ClaimActionType["execute"] = 4] = "execute";
    ClaimActionType[ClaimActionType["all"] = 5] = "all";
})(ClaimActionType = exports.ClaimActionType || (exports.ClaimActionType = {}));
