// @process-engine-js/data_model_contracts/src/entity_collection_reference.ts
import { IEntityCollectionReference } from '@process-engine-js/core_contracts';

export class EntityCollectionReference implements IEntityCollectionReference {

  private _namespace: string = undefined;
  private _type: string = undefined;
  private _relatedType: string = undefined;
  private _relatedAttribute: string = undefined;
  private _id: string = undefined;
  public resolve: boolean = false;

  constructor(namespace: string, type: string, id: string, relatedType: string, relatedAttribute: string, resolve?: boolean) {
    this._namespace = namespace;
    this._type = type;
    this._id = id;
    this._relatedType = relatedType;
    this._relatedAttribute = relatedAttribute;
    if (resolve) {
      this.resolve = resolve;
    }
  }

  public get namespace(): string {
    return this._namespace;
  }

  public get type(): string {
    return this._type;
  }

  public get id(): string {
    return this._id;
  }

  public get relatedType(): string {
    return this._relatedType;
  }

  public get relatedAttribute(): string {
    return this._relatedAttribute;
  }

  public toPojo(): any {
    const data: any = {
      id: this.id,
      _meta: {
        isCollectionRef: true,
        type: this.type,
        namespace: this.namespace,
        relatedType: this.relatedType,
        relatedAttribute: this.relatedAttribute
      }
    };

    return data;
  }
}
